pkg/arvo/app/chat-hook.hoon

```diff
@@ -0,0 +1,364 @@
+  ^-  (quip move _this)
+  ?-  -.diff
+      %keys    [~ this]
+      %config  [~ this]
+      %create  [~ this]
+      %read    [~ this]
+      %delete
+    ?.  (~(has by synced) path.diff)
+      [~ this]
+    :_  this(synced (~(del by synced) path.diff))
+    :-  (chat-poke diff)
+    [ost.bol %pull [%mailbox path.diff] [our.bol %chat-store] ~]~
+  ::
+      %message
+    :_  this
+    %+  turn  (prey:pubsub:userlib [%mailbox path.diff] bol)
+    |=  [=bone *]
+    ^-  move
+    [bone %diff [%chat-update diff]]
+  ::
+  ==
+::
+++  handle-foreign
+  |=  diff=chat-update
+  ^-  (quip move _this)
+  ?-  -.diff
+      %keys    [~ this]
+      %config  [~ this]
+      %read    [~ this]
+      %create
+    :_  this
+    ?>  ?=([* ^] path.diff)
+    =/  shp  (~(get by synced) path.diff)
+    ?~  shp  ~
+    ?.  =(u.shp src.bol)  ~
+    [(chat-poke [%create ship.diff t.path.diff])]~
+  ::
+      %delete
+    ?>  ?=([* ^] path.diff)
+    =/  shp  (~(get by synced) path.diff)
+    ?~  shp
+      [~ this]
+    ?.  =(u.shp src.bol)
+      [~ this]
+    :_  this(synced (~(del by synced) path.diff))
+    :-  (chat-poke diff)
+    [ost.bol %pull [%mailbox path.diff] [src.bol %chat-hook] ~]~
+  ::
+      %message
+    :_  this
+    ?>  ?=([* ^] path.diff)
+    =/  shp  (~(get by synced) path.diff)
+    ?~  shp  ~
+    ?.  =(src.bol u.shp)  ~
+    [(chat-poke diff)]~
+  ::
+  ==
+::
+++  quit
+  |=  wir=wire
+  ^-  (quip move _this)
+  ~&  chat-hook-quit+wir
+  ?:  =(wir /permissions)
+    :_  this
+    [ost.bol %peer /permissions [our.bol %permission-store] /updates]~
+  ?>  ?=([* ^] wir)
+  ?.  (~(has by synced) t.wir)
+    ::  no-op
+    [~ this]
+  ~&  %chat-hook-resubscribe
+  :_  (track-bone wir)
+  [ost.bol %peer wir [(slav %p i.t.wir) %chat-hook] wir]~
+::
+++  reap
+  |=  [wir=wire saw=(unit tang)]
+  ^-  (quip move _this)
+  ?~  saw
+    [~ this]
+  ?>  ?=(^ wir)
+  ~&  %chat-hook-reap
+  [((slog u.saw) ~) this(synced (~(del by synced) t.wir))]
+::
+++  chat-poke
+  |=  act=chat-action
+  ^-  move
+  [ost.bol %poke / [our.bol %chat-store] [%chat-action act]]
+::
+++  permission-poke
+  |=  act=permission-action
+  ^-  move
+  [ost.bol %poke / [our.bol %permission-store] [%permission-action act]]
+::
+++  create-permission
+  |=  [pax=path sec=chat-security]
+  ^-  (list move)
+  =/  read-perm   (weld pax /read)
+  =/  write-perm  (weld pax /write)
+  ?-  sec
+      %channel
+    :~  (permission-poke (sec-to-perm read-perm %black))
+        (permission-poke (sec-to-perm write-perm %black))
+    ==
+  ::
+      %village
+    :~  (permission-poke (sec-to-perm read-perm %white))
+        (permission-poke (sec-to-perm write-perm %white))
+    ==
+  ::
+      %journal
+    :~  (permission-poke (sec-to-perm read-perm %black))
+        (permission-poke (sec-to-perm write-perm %white))
+    ==
+  ::
+      %mailbox
+    :~  (permission-poke (sec-to-perm read-perm %white))
+        (permission-poke (sec-to-perm write-perm %black))
+    ==
+  ::
```

Comment:
Spacing comments should only ever be infix in "sub-hoons".

Reply:
Good point. Thanks.